### PR TITLE
Collect logs from cluster

### PR DESCRIPTION
**Pull request by aleksbykov**
- adding new command to sct - collect-logs
- adding new Class - LogCollector

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I gave variables/functions meaningful self-explanatory names
- [ ] I didn't leave commented-out/debugging code
- [ ] I didn't copy-paste code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (`hydra unit-tests`)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: sct.py [hunk 1]
```diff
@@ -308,5 +308,45 @@ def run_test(argv, backend, config, logdir):
                   failfast=False, buffer=False, catchbreak=True)
 
 
+@cli.command('collect-logs', help='Collect logs from cluster by test-id if test aborted unexpectedly')
```
Comment (bentsi) [inline sct.py]: no need to mention `aborted unexpectedly`

File: sdcm/logcollector.py [hunk 2]
```diff
@@ -0,0 +1,819 @@
+import datetime
+import io
+import logging
+import os
+import shutil
+# import sys
```
Comment (fruch) [inline sdcm/logcollector.py]: do leave commented imports here

File: sdcm/logcollector.py [hunk 3]
```diff
@@ -0,0 +1,819 @@
...
+import tempfile
+import time
+import requests
```
Comment (fruch) [inline sdcm/logcollector.py]: please separate third party imports from builtins  
see:
https://www.python.org/dev/peps/pep-0008/#imports

File: sdcm/logcollector.py [hunk 4]
```diff
@@ -0,0 +1,819 @@
...
+
+
+class Collector:
+    """Collector instance
+
+    Collector instance which should be run to collect logs and additional info
+    as separate stage in pipeline on as subcommand in hydra
+
+    """
+    LOCAL_SEARCH_FOR_USER = "abykov"
```
Comment (fruch) [inline sdcm/logcollector.py]: I don't understand what's this is for ? ... you are trying to identify if we are in JENKINS run or not ?

File: sdcm/logcollector.py [hunk 5]
```diff
@@ -0,0 +1,819 @@
...
+                      search_locally=True),
+        FileLogEntity(name='critical.log',
+                      search_locally=True)
+    ]
+    cluster_log_type = 'sct-runner'
+
+    def collect_logs(self, local_search_path=None):
+        for ent in self.log_entities:
+            ent.collect(None, self.local_dir, None, local_search_path=local_search_path)
+        # TODO: rewrite to zipfile module with zip_64
```
Comment (fruch) [inline sdcm/logcollector.py]: yep this is impotent since our files might get big

File: sdcm/logcollector.py [hunk 6]
```diff
@@ -0,0 +1,819 @@
...
+        as single stage in pipeline for defined cluster sets
+        """
+        results = {}
+        if not self.test_id:
+            self.get_current_test_id()
+            LOGGER.info("Found latest test_id: {0.test_id}".format(self))
+
+        self.get_running_cluster_sets_by_testid()
+
+        local_dir_with_logs = self.get_local_testdir()
```
Comment (fruch) [inline sdcm/logcollector.py]: if this return None, the whole thing won't work ? or do we a have a fallback here to do part of the work into a temp directory

I would love to see this working with logging into the jenkins server, i.e. collect can we can in case we are not in jenkins. i.e. db logs can be collected directly from journalctl. yes job logs and events log would be missing, but in some cases it's not that important.

File: sdcm/logcollector.py [hunk 7]
```diff
@@ -0,0 +1,819 @@
...
+            return self.local_dir, None, None
+
+        def collect_logs_per_node(node):
+            remote_node_dir = self.create_remote_storing_dir(node)
+            local_node_dir = os.path.join(self.local_dir, node.name)
+            os.makedirs(local_node_dir)
+            for log_entity in self.log_entities:
+                log_entity.collect(node, local_node_dir, remote_node_dir, local_search_path=local_search_path)
+
+        ParallelObject(self.nodes, num_workers=3, timeout=300).run(collect_logs_per_node)
```
Comment (fruch) [inline sdcm/logcollector.py]: why num_workers is only 3 ? what if we have 10 nodes ? or more ? would it work, or just work slower ?

File: sdcm/logcollector.py [hunk 8]
```diff
@@ -0,0 +1,819 @@
...
+            phantomjs_base = self.phantomjs_base
+            phantomjs_tar = "{phantomjs_base}.tar.bz2".format(**locals())
+            phantomjs_url = "https://bitbucket.org/ariya/phantomjs/downloads/{phantomjs_tar}".format(**locals())
+            install_phantom_js_script = dedent("""
+                rm -rf {phantomjs_base}*
+                curl {phantomjs_url} -o {phantomjs_tar} -L
+                tar xvfj {phantomjs_tar}
+            """.format(**locals()))
+            localrunner.run("bash -ce '%s'" % install_phantom_js_script)
+            localrunner.run("cd phantomjs-2.1.1-linux-x86_64 && sed -e 's/200);/10000);/' examples/rasterize.js |grep -v 'use strict' > r.js")
```
Comment (fruch) [inline sdcm/logcollector.py]: use `phantomjs_base`

File: sdcm/logcollector.py [hunk 9]
```diff
@@ -0,0 +1,819 @@
...
+    """Collect Grafana screenshot
+
+    Collect grafana screenshot
+
+    Extends:
+        GrafanaEntity
+    """
+
+    def _get_screenshot_link(self, grafana_url, screenshot_path):
+        LocalCmdRunner().run("cd phantomjs-2.1.1-linux-x86_64 && bin/phantomjs r.js \"%s\" \"%s\" 1920px" % (
```
Comment (fruch) [inline sdcm/logcollector.py]: use `phantomjs_base`

File: sdcm/logcollector.py [hunk 10]
```diff
@@ -0,0 +1,819 @@
...
+    Extends:
+        GrafanaEntity
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(GrafanaSnapshotEntity, self).__init__(*args, **kwargs)
+
+    def _get_shared_snapshot_link(self, grafana_url):
+        result = LocalCmdRunner().run(
+            "cd phantomjs-2.1.1-linux-x86_64 && bin/phantomjs ../data_dir/share_snapshot.js \"%s\"" % (grafana_url))
```
Comment (fruch) [inline sdcm/logcollector.py]: use `phantomjs_base`

File: sdcm/logcollector.py [hunk 11]
```diff
@@ -0,0 +1,819 @@
...
+
+from textwrap import dedent
+
+from utils.common import (S3Storage, list_instances_aws, list_instances_gce,
+                          retrying, ParallelObject)
+from .db_stats import PrometheusDBStats
+from .remote import LocalCmdRunner, RemoteCmdRunner
+
+LOGGER = logging.getLogger(__name__)
+LOGGER.setLevel(logging.INFO)
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/logcollector.py]: removed

File: sdcm/cluster.py [hunk 12]
```diff
@@ -3657,71 +3669,81 @@ def stop_scylla_monitoring(self, node):
...
+    #         localrunner.run("cd phantomjs-2.1.1-linux-x86_64 && sed -e 's/200);/10000);/' examples/rasterize.js |grep -v 'use strict' > r.js")
+    #         self.phantomjs_installed = True
+    #     else:
+    #         self.log.debug("PhantomJS is already installed!")
 
     def get_grafana_screenshot_and_snapshot(self, test_start_time=None):
         """
             Take screenshot of the Grafana per-server-metrics dashboard and upload to S3
         """
+        from logcollector import GrafanaSnapshotEntity, GrafanaScreenShotEntity
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/cluster.py]: this required for method update_test_details
this will be removed from teardown

File: sdcm/logcollector.py (hunk 4 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: Yes, you right. Just forgot to change to jenkins

File: sdcm/logcollector.py (hunk 5 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: Done

File: sdcm/logcollector.py (hunk 10 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: fixed

File: sdcm/logcollector.py (hunk 9 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: fixed

File: sdcm/logcollector.py (hunk 8 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: fixed

File: sdcm/logcollector.py (hunk 7 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: will work slowly. We can use : int(num of nodes / 2) to prevent overload if, for example, there are 500 nodes?

File: sdcm/logcollector.py (hunk 6 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: if get_local_testdir will return none, then the new dir will be created in $USER/sct_results folder
and all logs (except sct logs (critical.log, events.log, sct.logs) will be collected (db_cluster, monitor) also the test_id will be written to this new folder, and any future log collecting will stored in same folder
if run collect logs from local host, all logs will colleted from remote hosts (db, monitor, loader).
In next phase, there is an idea to find appropriate folder by test id on builders and collect sct logs.

Comment (fruch) [reply sdcm/logcollector.py]: Can we do this a bit more explicit and not based on the username, for example if I'll run it local on my desk I just want to pass the test-id and it to work, not just when calling it on the worker.

also you'll aws credentials to run hydra commands, I'm not sure it would be available on all workers

File: sdcm/logcollector.py (hunk 4 above)
Comment (fruch) [reply sdcm/logcollector.py]: I think we want the to work from everywhere and not just from Jenkins workers. 

We should identify the worker, and copy needed files out of it, workers can be tagged in a way we could list them, and we can scan them for the specific test id

File: vars/newtestpipeline.groovy [hunk 13]
```diff
@@ -0,0 +1,127 @@
+#!groovy
```
Comment (fruch) [inline vars/newtestpipeline.groovy]: please don't duplicate this file, please add it to the longevity one.
it's safe enough as long you are on your branch testing it.

File: sdcm/utils/common.py [hunk 14]
```diff
@@ -373,141 +369,40 @@ def get_latest_gemini_version():
...
         return results
 
+    def convert_to_date(date_str):
+        try:
+            t = datetime.datetime.strptime(date_str, "%Y%m%d_%H%M%S")
+        except ValueError:
+            try:
+                t = datetime.datetime.strptime(date_str, "%Y_%m_%d_%H_%M_%S")
+            except ValueError:
+                t = datetime.datetime(1999, 01, 01, 01, 01, 01)
```
Comment (fruch) [inline sdcm/utils/common.py]: why are you prefixing all with zero ?

File: vars/newtestpipeline.groovy [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+                                set -xe
+                                env
+
+                                export SCT_CLUSTER_BACKEND="${params.backend}"
+                                export SCT_REGION_NAME=${aws_region}
+                                export SCT_CONFIG_FILES="${test_config}"
+
+                                echo "start collect logs ..."
+                                ./docker/env/hydra.sh collect-logs --logdir /sct
```
Comment (fruch) [inline vars/newtestpipeline.groovy]: please add a step for cleaning the resources after the log are collected
I guess it should be a new step, or something, it should happen even if logs are failed to be collected

File: sdcm/cluster.py [hunk 16]
```diff
@@ -1123,7 +1134,7 @@ def _report_housekeeping_uuid(self, verbose=False):
                                        ignore_status=True, verbose=verbose)
         if uuid_result.exit_status == 0 and mark_result.exit_status != 0:
             result = self.remoter.run('cat %s' % uuid_path, verbose=verbose)
-            self.remoter.run(cmd % result.stdout.strip(), ignore_status=True)
+            self.remoter.run(cmd % result.stdout.strip())
```
Comment (fruch) [inline sdcm/cluster.py]: why are you touch this ?, please remove

File: sdcm/cluster.py [hunk 17]
```diff
@@ -1479,7 +1490,6 @@ def download_scylla_repo(self, scylla_repo):
             result = self.remoter.run('cat %s' % repo_path, verbose=True)
             verify_scylla_repo_file(result.stdout, is_rhel_like=True)
             self.remoter.run('sudo yum clean all')
-            self.remoter.run('sudo rm -rf /var/cache/yum/*')
```
Comment (fruch) [inline sdcm/cluster.py]: please don't touch other logic

File: sdcm/cluster.py [hunk 18]
```diff
@@ -1504,7 +1514,6 @@ def clean_scylla(self):
         if self.is_rhel_like():
             self.remoter.run('sudo yum remove -y scylla\*')
             self.remoter.run('sudo yum clean all')
-            self.remoter.run('sudo rm -rf /var/cache/yum/*')
```
Comment (fruch) [inline sdcm/cluster.py]: please don't touch other logic

File: sdcm/cluster.py [hunk 19]
```diff
@@ -3275,7 +3283,7 @@ def __init__(self, targets, params):
         self.local_metrics_addr = start_metrics_server()  # start prometheus metrics server locally and return local ip
         self.sct_ip_port = self.set_local_sct_ip()
         self.grafana_port = 3000
-        self.monitor_branch = self.params.get('monitor_branch')
+        self.monitor_branch = self.params.get('monitor_branch', default='branch-2.4')
```
Comment (fruch) [inline sdcm/cluster.py]: please don't touch other logic

File: sdcm/logcollector.py [hunk 20]
```diff
@@ -0,0 +1,849 @@
...
+
+
+class Collector:
+    """Collector instance
+
+    Collector instance which should be run to collect logs and additional info
+    as separate stage in pipeline on as subcommand in hydra
+
+    """
+    LOCAL_SEARCH_FOR_USER = "jenkins"
```
Comment (fruch) [inline sdcm/logcollector.py]: I still don't understand what would happen if I run this on my computer after I've run a test:
```
hydra collect-logs `cat ~/sct-results/latest/test_id`
```

And what would happen if I've run from my computer 
```
# where test_id is a from a job running in jenkins
hydra collect-logs [test_id]
```

File: vars/newtestpipeline.groovy (hunk 15 above)
Comment (aleksbykov) [reply vars/newtestpipeline.groovy]: added

File: sct.py [hunk 21]
```diff
@@ -362,5 +369,168 @@ def cloud_usage_report(emails):
...
+    for cluster_type, s3_link in collected_logs.items():
+        table.add_row([cluster_type, s3_link])
+    click.echo(table.get_string(title="Collected logs by test-id: {} Storing dir: {}".format(collector.test_id,
+                                                                                             collector.storing_dir,)))
+
+
+@cli.command('destroy-resources', help='')
+@click.option('--logdir', type=str)
+@click.option('--backend', type=str)
+# pylint: disable=too-many-statements
```
Comment (fruch) [inline sct.py]: please put those comment inside the function, here they affect the rest of the file

File: sct.py [hunk 22]
```diff
@@ -362,5 +369,168 @@ def cloud_usage_report(emails):
...
+# pylint: disable=too-many-branches
+def destroy_resources(logdir, backend):
```
Comment (fruch) [inline sct.py]: what's the difference between destory-resources and clean-resources ? or is should be just a rename ?

File: sct.py [hunk 23]
```diff
@@ -362,5 +369,168 @@ def cloud_usage_report(emails):
...
+
+            if config['post_behavior_loader_nodes'] == 'keep':
+                LOGGER.info('Leave instances runing')
+
+            if config['post_behavior_loader_nodes'] == 'keep-on-failure':
+                LOGGER.info('Checking status of run in ES...')
+                if status:
+                    LOGGER.info('Run failed. Leave instances running')
+                else:
+                    LOGGER.info('Run wasSuccessful. kill nodes')
```
Comment (fruch) [inline sct.py]: missing a space in `wasSuccessful`

File: sct.py (hunk 22 above)
Comment (fruch) [reply sct.py]: also this function has lot of logic, I would suggest breaking it into multiple functions, maybe on per backend, and take it out of sct.py, think of sct.py as the UI, should have too much logic in it.

File: sdcm/sct_config.py [hunk 24]
```diff
@@ -819,6 +819,21 @@ class SCTConfiguration(dict):
...
         dict(name="logs_transport", env="SCT_LOGS_TRANSPORT", type=str,
              help="How to transport logs: rsyslog or ssh"),
+
+        dict(name="collect_logs", env="SCT_COLLECT_LOGS", type=boolean,
+             help="Collect logs from instances and sct runner"),
+
+        dict(name="execute_post_behavior", env="SCT_EXECUTE_POST_BEHAVIOR", type=boolean,
+             help="Run post behavior actions in sct teardown step"),
+
+        dict(name="post_behavior_db_nodes", env="SCT_POST_BEHAVIOR_DB_NODES", type=str,
```
Comment (fruch) [inline sdcm/sct_config.py]: if we are introducing a those new options, what about the old `failure_post_behavior` ? it would be removed ? or automatically mapped to those ?

File: vars/newtestpipeline.groovy [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+                                    def aws_region = groovy.json.JsonOutput.toJson(params.aws_region)
+                                    def test_config = groovy.json.JsonOutput.toJson(pipelineParams.test_config)
+
+                                    sh """
+                                    #!/bin/bash
+
+                                    set -xe
+                                    env
+
+                                    export SCT_CLUSTER_BACKEND="${params.backend}"
```
Comment (fruch) [inline vars/newtestpipeline.groovy]: since those are command we want to be able to run manually, please try avoid using the environment variable and pass everything need via the command line arguments

Comment (fruch) [reply vars/newtestpipeline.groovy]: for example I don't think the configuration file is needed here

File: sdcm/cluster.py [hunk 26]
```diff
@@ -174,8 +172,10 @@ class Setup(object):
     BACKTRACE_DECODING = False
     INTRA_NODE_COMM_PUBLIC = False
     RSYSLOG_ADDRESS = None
+    ES_DOC_ID = None
```
Comment (fruch) [inline sdcm/cluster.py]: where this one is used ?

File: sdcm/cluster.py [hunk 27]
```diff
@@ -184,6 +184,20 @@ class Setup(object):
     def test_id(cls):
         return cls._test_id
 
+    @classmethod
```
Comment (fruch) [inline sdcm/cluster.py]: might want to turn this into a property

File: sdcm/cluster.py [hunk 28]
```diff
@@ -1751,12 +1751,12 @@ def stop_scylla_manager_log_capture(self, timeout=10):
         self._scylla_manager_journal_thread.join(timeout)
         self._scylla_manager_journal_thread = None
 
-    def collect_mgmt_log(self):
-        self.log.debug("Collect scylla manager log ...")
-        mgmt_log_name = "/tmp/{}_scylla_manager.log".format(self.name)
-        self.remoter.run('sudo journalctl -u scylla-manager.service --no-tail > {}'.format(mgmt_log_name), ignore_status=True, verbose=False)
-        self.log.debug("Collected log : {}".format(mgmt_log_name))
-        return mgmt_log_name
+    # def collect_mgmt_log(self):
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/cluster.py]: removed

File: sdcm/logcollector.py (hunk 2 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: removed

File: sdcm/logcollector.py (hunk 4 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: Adding scan for test id on builders. but list of builders is hardcoded.

File: sdcm/logcollector.py (hunk 6 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: Fixed. Now first search on local host ( if it run manually or by builder) and then search the running instances and collect from them. For sct logs, if not found locally, will try to find on builders.

File: sdcm/logcollector.py (hunk 7 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: implemented

File: vars/newtestpipeline.groovy (hunk 13 above)
Comment (aleksbykov) [reply vars/newtestpipeline.groovy]: fixed

File: sdcm/utils/common.py (hunk 14 above)
Comment (aleksbykov) [reply sdcm/utils/common.py]: fixed

File: sdcm/cluster.py (hunk 19 above)
Comment (aleksbykov) [reply sdcm/cluster.py]: done

File: sdcm/logcollector.py (hunk 20 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: implemented. new command return builder and folder of testrun by test-id.
limitation is that builders are hardcoded

File: sct.py (hunk 21 above)
Comment (aleksbykov) [reply sct.py]: done

File: sct.py (hunk 22 above)
Comment (aleksbykov) [reply sct.py]: refactored and moved to clean-resources. but now clean-resources looks a bit complex

File: sct.py (hunk 23 above)
Comment (aleksbykov) [reply sct.py]: fixed

File: sdcm/sct_config.py (hunk 24 above)
Comment (aleksbykov) [reply sdcm/sct_config.py]: i think old one should be totally replaced with new ones
Current design, if new parameters are not in config file, then old config is used.

File: vars/newtestpipeline.groovy (hunk 25 above)
Comment (aleksbykov) [reply vars/newtestpipeline.groovy]: adding explicit parameters for command clean-resources, collect-logs.

File: sdcm/cluster.py (hunk 27 above)
Comment (aleksbykov) [reply sdcm/cluster.py]: it should be removed once the test_id generated by hydra, will become the document_id in ElasticSearch
think, behavior should be same as for test_id

File: sdcm/cluster.py [hunk 29]
```diff
@@ -2901,26 +2914,25 @@ def restart_scylla(self, nodes=None):
...
-                pkg_list_file.write(node.get_installed_packages())
-            with open(os.path.join(storing_dir_for_node, 'console_output'), 'w') as console_output_file:
-                console_output_file.write(node.get_console_output())
-            with open(os.path.join(storing_dir_for_node, 'console_screenshot.jpg'), 'wb') as cscrn:
-                imagedata = node.get_console_screenshot()
-                cscrn.write(decodestring(imagedata))
-            if os.path.exists(node.database_log):
-                shutil.copy(node.database_log, storing_dir_for_node)
-        return storing_dir
+    # def collect_logs(self, storing_dir):
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/cluster.py]: removed

**Comment by aleksbykov** [comment]
Example of job:
http://jenkins.scylladb.com/view/staging/job/scylla-staging/job/longevity-abykov-staging/job/longevity-1tb-7days-2Disruptive-Nemesises/

File: sdcm/sct_config.py [hunk 30]
```diff
@@ -824,6 +824,21 @@ class SCTConfiguration(dict):
...
+             help=""),
+
+        dict(name="post_behavior_loader_nodes", env="SCT_POST_BEHAVIOR_DB_NODES", type=str,
```
Comment (bentsi) [inline sdcm/sct_config.py]: SCT_POST_BEHAVIOR_**LOADER**_NODES

File: sdcm/sct_config.py [hunk 31]
```diff
@@ -824,6 +824,21 @@ class SCTConfiguration(dict):
...
+             help=""),
+
+        dict(name="post_behavior_monitor_nodes", env="SCT_POST_BEHAVIOR_DB_NODES", type=str,
```
Comment (bentsi) [inline sdcm/sct_config.py]: SCT_POST_BEHAVIOR_**MONITOR**_NODES

File: sdcm/tester.py [hunk 32]
```diff
@@ -1396,36 +1399,37 @@ def stop_resources(self):
...
-                self.monitors.destroy()
-                self.monitors = None
-            elif self._failure_post_behavior == 'stop':
-                for node in self.monitors.nodes:
-                    node.instance.stop()
-                self.monitors = None
+        if self.params.get('execute_post_behavior', False):
+            self.log.debug('Cleaning up resources used in the test')
+
+        # if self.db_cluster is not None:
```
Comment (bentsi) [inline sdcm/tester.py]: why is the commented code here?

File: sdcm/tester.py [hunk 33]
```diff
@@ -1450,60 +1454,18 @@ def tearDown(self):
...
-            remove_files(job_log_dir_tmp)
-            remove_files(joblog_archive)
-        self.log.info('Test ID: {}'.format(cluster.Setup.test_id()))
-        if self.create_stats:
-            self.log.info("ES document id: {}".format(self.get_doc_id()))
-
-    def collect_events_log(self):
-        event_log_dir = EVENTS_PROCESSES['MainDevice'].event_log_base_dir
+            if self.params.get('collect_logs', False):
+                storing_dir = os.path.join(self.logdir, "collected_logs")
```
Comment (bentsi) [inline sdcm/tester.py]: storage_dir

File: sdcm/utils/monitorstack.py [hunk 34]
```diff
@@ -0,0 +1,251 @@
...
+                   archive=archive,
+                   archive_name=os.path.basename(archive)))
+    result = LocalCmdRunner().run(cmd, ignore_status=True)
+    if result.exited > 0:
+        LOGGER.error("Error during extracting prometheus snapshot. Switch to next archive")
+        return False
+    return get_monitor_data_dir(monitor_data_base_dir)
+
+
+def create_monitoring_stack_folder(base_dir, archive):
```
Comment (bentsi) [inline sdcm/utils/monitorstack.py]: folders are in MS Windows, in Linux they are called directories
 `create_monitoring_stack_folder` -> `create_monitoring_stack_dir`

File: sdcm/utils/monitorstack.py [hunk 35]
```diff
@@ -0,0 +1,251 @@
...
+            LOGGER.warning('Monitor stack archive for date %s was not found', date_time)
+        else:
+            arch = found_archives[0]
+    else:
+        arch = archives[0]
+
+    return arch
+
+
+def create_monitoring_data_folder(base_dir, archive):
```
Comment (bentsi) [inline sdcm/utils/monitorstack.py]: create_monitoring_data_**dir**

File: test-cases/longevity/longevity-10gb-3h.yaml [hunk 36]
```diff
@@ -1,5 +1,5 @@
-test_duration: 180
-stress_cmd: ["cassandra-stress write cl=QUORUM duration=180m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=1000 -pop seq=1..10000000 -log interval=5"
+test_duration: 10
```
Comment (bentsi) [inline test-cases/longevity/longevity-10gb-3h.yaml]: why did you change this file ?

File: data_dir/share_snapshot.js [hunk 37]
```diff
@@ -10,39 +10,39 @@ page.viewportSize = { width: 1920, height: 1080 };
...
-		phantom.exit();
-	}
-	window.setTimeout(function () {
-		page.evaluate(function() {
-			var ev = document.createEvent('MouseEvent');
-			ev.initMouseEvent('click', true, true);
-			var l = document.getElementsByClassName("navbar-button--share");
-        	l[0].dispatchEvent(ev)
-		});
+    if (status !== "success") {
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply data_dir/share_snapshot.js]: done

File: test-cases/longevity/longevity-10gb-3h.yaml (hunk 36 above)
Comment (aleksbykov) [reply test-cases/longevity/longevity-10gb-3h.yaml]: it was changed for testing purposes. Revert it back. Adding new config parameters

File: sdcm/utils/monitorstack.py (hunk 34 above)
Comment (aleksbykov) [reply sdcm/utils/monitorstack.py]: renamed

File: sdcm/tester.py (hunk 33 above)
Comment (aleksbykov) [reply sdcm/tester.py]: Renamed everywhere

File: sdcm/sct_config.py (hunk 31 above)
Comment (aleksbykov) [reply sdcm/sct_config.py]: fixed

File: sdcm/sct_config.py (hunk 30 above)
Comment (aleksbykov) [reply sdcm/sct_config.py]: fixed

File: sdcm/utils/common.py [hunk 38]
```diff
@@ -1085,3 +972,224 @@ def update_certificates():
...
+]
+
+
+def get_builder_by_test_id(test_id):
+    from sdcm.remote import RemoteCmdRunner
+
+    base_path_on_builder = "/home/jenkins/slave/workspace"
+    found_builders = []
+
+    def search_builder(builder):
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/utils/common.py]: fixed

**Comment by roydahan** [comment]
@aleksbykov you have conflicts here.
Also not all comments addressed.

File: sdcm/tester.py (hunk 32 above)
Comment (aleksbykov) [reply sdcm/tester.py]: Rewrote this code to support new post_behavior and allow to control clean instances for manual running sct from local hosts

File: sdcm/utils/common.py [hunk 39]
```diff
@@ -1085,3 +972,224 @@ def update_certificates():
...
+
+        if not result.exited and not result.stderr:
+            path = result.stdout.strip()
+            LOGGER.info("Builder name %s, ip %s, folder %s", builder['name'], builder['public_ip'], path)
+            return {"builder": builder, "path": os.path.dirname(path)}
+        else:
+            LOGGER.info("Nothing found")
+            return None
+
+    search_obj = ParallelObject(BUILDERS, timeout=30, num_workers=int(len(BUILDERS) / 2))
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/utils/common.py]: to decrease the number of concurrent ssh connections to large number  of builders from localhost.
Set to len(list_of_builder)

File: sdcm/cluster.py [hunk 40]
```diff
@@ -3705,49 +3674,15 @@ def get_grafana_screenshot_and_snapshot(self, test_start_time=None):  # pylint:
...
-                    snapshots.append(self._get_shared_snapshot_link(grafana_url))
-
-                return {'screenshots': screenshots, 'snapshots': snapshots}
-
-        except Exception as details:  # pylint: disable=broad-except
-            self.log.error('Error taking monitor snapshot: %s',  # pylint: disable=no-member
-                           str(details))
-            return {}
+        for node in self.nodes:  # pylint: disable=no-member
+            screenshot_collector = GrafanaScreenShotEntity(name="grafana-screenshot",
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/cluster.py]: Renamed

File: sct.py [hunk 41]
```diff
@@ -70,26 +74,60 @@ def provision(**kwargs):
...
+
+    else:
+        if not (user or test_id):
+            click.echo(clean_resources.get_help(ctx))
+            return
+
+        if user:
+            params['RunByUser'] = user
+
+        if not test_id:
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sct.py]: got what you mean. Fixed

**Comment by bentsi** [comment]
please resolve conflicts

File: sdcm/tester.py [hunk 42]
```diff
@@ -1463,60 +1477,18 @@ def tearDown(self):
...
         finally:
-            if self._failure_post_behavior == 'destroy':
-                clean_cloud_instances({"TestId": str(cluster.Setup.test_id())})
             stop_events_device()
-            self.collect_events_log()
-            self.zip_and_upload_job_log()
-            stop_rsyslog()
-
-    def zip_and_upload_job_log(self):
+            if self.params.get('collect_logs', False):
```
Comment (bentsi) [inline sdcm/tester.py]: remove the default, it should be set in defaults yaml

File: defaults/test_default.yaml [hunk 43]
```diff
@@ -29,10 +29,14 @@ nemesis_filter_seeds: true
...
+post_behavior_db_nodes: keep-on-failure
+post_behavior_loader_nodes: destroy
+post_behavior_monitor_nodes: keep
 
 cloud_credentials_path: '~/.ssh/support'
 
 backtrace_decoding: false
 
 logs_transport: "rsyslog"
+collect_logs: true
```
Comment (bentsi) [inline defaults/test_default.yaml]: default should be `false`, we we collect logs in Jenkins pipeline (and if somebody wants it in his private test YAML should enable it explicitly)

Comment (amoskong) [reply defaults/test_default.yaml]: You can't assume some log isn't necessary in analyzing failed jobs. So please save all logs generated by SCT. It's general requirement, not just for private testing.

Comment (aleksbykov) [reply defaults/test_default.yaml]: Set to false.
All logs are collected as before. This parameter only disable collect logs in teardown method. 
All logs will be collected by new stage in pipeline.

File: sdcm/tester.py (hunk 42 above)
Comment (aleksbykov) [reply sdcm/tester.py]: done

File: sdcm/logcollector.py (hunk 3 above)
Comment (aleksbykov) [reply sdcm/logcollector.py]: fixed

File: sdcm/cluster.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (aleksbykov) [reply sdcm/cluster.py]: fixed

File: vars/longevityPipeline.groovy [hunk 44]
```diff
@@ -55,45 +55,105 @@ def call(Map pipelineParams) {
...
+                                    ./docker/env/hydra.sh collect-logs --logdir /sct
+                                    echo "end collect logs"
+                                    """
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            stage('Clear instances') {
```
Comment (bentsi) [inline vars/longevityPipeline.groovy]: Clean resources

Comment (aleksbykov) [reply vars/longevityPipeline.groovy]: fixed
